mdata/cache/ccache_metric.go

```diff
@@ -69,6 +65,125 @@ func (mc *CCacheMetric) Del(ts uint32) int {
 	return len(mc.chunks)
 }
 
+// AddRange adds a range (sequence) of chunks.
+// Note the following requirements:
+// the sequence should be in ascending timestamp order
+// the sequence should be complete (no gaps)
+func (mc *CCacheMetric) AddRange(prev uint32, itergens []chunk.IterGen) {
+	if len(itergens) == 0 {
+		return
+	}
+
+	if len(itergens) == 1 {
+		mc.Add(prev, itergens[0])
+		return
+	}
+
+	mc.Lock()
+	defer mc.Unlock()
+
+	// pre-allocate 1 slice, cheaper than allocating one by one
+	chunks := make([]CCacheChunk, 0, len(itergens))
+
+	// handle the first one
+	itergen := itergens[0]
+	ts := itergen.Ts
+
+	addKeysDirect := len(mc.keys) == 0 || mc.keys[len(mc.keys)-1] < ts
+
+	// if previous chunk has not been passed we try to be smart and figure it out.
+	// this is common in a scenario where a metric continuously gets queried
+	// for a range that starts less than one chunkspan before now().
+	if prev == 0 {
+		res, ok := mc.seekDesc(ts - 1)
+		if ok {
+			prev = res
+		}
+	}
+
+	// if the previous chunk is cached, link it
+	if _, ok := mc.chunks[prev]; ok {
+		mc.chunks[prev].Next = ts
+	} else {
+		prev = 0
+	}
+
+	// add chunk if we don't have it yet (most likely)
+	if _, ok := mc.chunks[ts]; !ok {
+		chunks = append(chunks, CCacheChunk{
+			Ts:    ts,
+			Prev:  prev,
+			Next:  itergens[1].Ts,
+			Itgen: itergen,
+		})
+		mc.chunks[ts] = &chunks[len(chunks)-1]
+		if addKeysDirect {
+			mc.keys = append(mc.keys, ts)
+		}
+	}
+
+	prev = ts
+
+	// handle the 2nd until the last-but-one
+	for i := 1; i < len(itergens)-1; i++ {
+		itergen := itergens[i]
+		ts := itergen.Ts
+		// add chunk if we don't have it yet (most likely)
+		if _, ok := mc.chunks[ts]; !ok {
+			chunks = append(chunks, CCacheChunk{
+				Ts:    ts,
+				Prev:  prev,
+				Next:  itergens[i+1].Ts,
+				Itgen: itergen,
+			})
+			mc.chunks[ts] = &chunks[len(chunks)-1]
+			if addKeysDirect {
+				mc.keys = append(mc.keys, ts)
+			}
+		}
+		prev = ts
+	}
+
+	// handle the last one
+	itergen = itergens[len(itergens)-1]
+	ts = itergen.Ts
+
+	// if nextTs() can't figure out the end date it returns ts
+	next := mc.nextTsCore(itergen, ts, prev, 0)
```

Comment:
if the chunk at `ts` is already cached (the `ok` on line `165` is `true`), then there is no need to figure out what `next` should be. So the block `152-162` could be moved down into that if block